.github/workflows/ci.yml

```diff
@@ -197,3 +213,47 @@ jobs:
         with:
           files: .
           config_file: .markdownlint.yaml
+
+  python-checks:
+    timeout-minutes: 5
+    runs-on: ubuntu-22.04
+    name: python-${{ matrix.tool }}
+    strategy:
```

Comment:
Why you use matrix here? Looks like in one job it should work faster and much simpler for maintaining.

Reply:
Matrix is just a tool to get **one GitHub Action job per check**. Why? It seems to fit better for the independent nature of the checks:
- checks better be run concurrently, alongside each other
- checks should run independently: if one fails, other should continue
- checks have independent results that would be nice to see reported independently (in realtime, as we got them), at a glance:

![image](https://github.com/ryukzak/nitta/assets/5229130/02734720-c053-44f4-bd08-69ae169ec317)

On the other hand, the "one job" approach you suggest forces us to use **job steps** for each check:
- Job steps are designed to be sequential - prevents concurrent running.
- While steps support `continue-on-error`, checks like that seem not to be the primary use case for that. I didn't test myself, but someone on SO said you even get a green checkmark for a failed step with `continue-on-error`. Sounds awful.

Yes, running checks concurrently wastes some GitHub Actions time on multiple initializations, but it happens at the same time on multiple machines, so it's the same 20 seconds of your time. Since GitHub Actions time is free for this project, I don't see it as a concern :)

More than that, in the end it saves your time: while mypy is still running for another **couple of minutes** (wow, we should save the .mypy_cache, perhaps), you can already see whether the ruff or vulture failed (right away as the check has finished) and work on it.

For those reasons I actually have a quite firm opinion here that multiple jobs are a better fit DX-wise.

Reply:
So, back to the matrix: initially, there was none. I wrote a bunch of copy-pased jobs, realized they are very similar and tried to find a way to keep them DRY. Pretty happy with the result, in fact. It's literally a way to parametrize similar job definitions, much like a function in programming allows to reuse similar pieces of code.

They may require a bit more effort to extend the behaviour (like I did with mypy and `setup_cmd`), but it's still pretty straightforward once you grasp the features at your disposal. Consider this: now we've realized the need to cache the `.mypy_cache`, we implement it and "look, we get the saving of `.ruff_cache` as a gift, for free"! Cool, isn't it?